File: plugins/Dolly/Dolly.cpp
// Dolly.cpp : Defines the initialization routines for the DLL.
//

#include "stdafx.h"
#include "Dolly.h"
#include <chrono>

PLUGIN_API
AM_RESPONSE am_Msg(AM_MSG msg, AM_UINT wParam, AM_INT lParam)
{
  switch( msg )
  {
  case AM_MSG_INIT:
    {      
      //
      // the plugin manager.
      AmPlugin* p = (AmPlugin*)(lParam);

      //
      // get the full path of this plugin
      // call 'getCommand( 0, ... ) and it will return this file.
      WCHAR szPath[MAX_PATH];
      size_t l = p->GetCommand( 0, MAX_PATH, szPath );

      //
      // add an actio called 'dolly'
      p->AddAction( L"Hello, Dolly!", szPath );
    }
    break;

  case AM_MSG_DEINIT:
    // clean up
    break;

  case AM_MSG_MAIN:
    {
      //  seed me up...
      const auto seed = std::chrono::system_clock::now().time_since_epoch().count();
      std::srand( static_cast<unsigned int>(seed));

      // just say something.
      const
      WCHAR* lyrics[] = {L"Hello, Dolly",
                        L"Well, hello, Dolly",
                        L"It's so nice to have you back where you belong",
                        L"You're lookin' swell, Dolly",
                        L"I can tell, Dolly",
                        L"You're still glowin', you're still crowin'",
                        L"You're still goin' strong",
                        L"We feel the room swayin'",
                        L"While the band's playin'",
                        L"One of your old favourite songs from way back when",
                        L"So, take her wrap, fellas",
                        L"Find her an empty lap, fellas",
                        L"Dolly'll never go away again",
                        L"Hello, Dolly",
                        L"Well, hello, Dolly",
                        L"It's so nice to have you back where you belong",
                        L"You're lookin' swell, Dolly",
                        L"I can tell, Dolly",
                        L"You're still glowin', you're still crowin'",
                        L"You're still goin' strong",
                        L"We feel the room swayin'",
                        L"While the band's playin'",
                        L"One of your old favourite songs from way back when",
                        L"Golly, gee, fellas",
                        L"Find her a vacant knee, fellas",
                        L"Dolly'll never go away",
                        L"Dolly'll never go away",
                        L"Dolly'll never go away again",
      };

      // the range of numbers.
      const double range_max = _countof(lyrics);

      // get the random number
      const double rnd = static_cast<double>( std::rand());

      // make sure it is in range.
      const int u = static_cast<int>( rnd / (RAND_MAX + 1) * (range_max));

      const WCHAR* lyric = lyrics[u];

      // and then just say it...
      ((AmPlugin*)(lParam))->Say( lyric, 200, 5 );

      //  all good.
      return AM_RESP_SUCCESS;
    }
    break;

  default:
    break;
  }

  // not handled here.
  return AM_RESP_NOT_SUPPORTED;
}

File: plugins/Dolly.NET/Dolly.cs
//This file is part of Myoddweb.Piger.
//
//    Myoddweb.Piger is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    Myoddweb.Piger is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with Myoddweb.Piger.  If not, see<https://www.gnu.org/licenses/gpl-3.0.en.html>.
using System;
using System.Threading;
using System.Threading.Tasks;
using ActionMonitor.Interfaces;

namespace Dolly.NET
{
  public class Dolly : IAction
  {
    /// <summary>
    /// The action monitor instance.
    /// </summary>
    private IActionMonitor _monitor;

    /// <summary>
    /// A random item so we can pick a sentence
    /// </summary>
    private readonly Random _random = new Random();

    public Task<bool> GoAsync(CancellationToken token)
    {
      var sentences = new[]
      {
        "Hello, Dolly",
        "Well, hello, Dolly",
        "It's so nice to have you back where you belong",
        "You're lookin' swell, Dolly",
        "I can tell, Dolly",
        "You're still glowin', you're still crowin'",
        "You're still goin' strong",
        "We feel the room swayin'",
        "While the band's playin'",
        "One of your old favourite songs from way back when",
        "So, take her wrap, fellas",
        "Find her an empty lap, fellas",
        "Dolly'll never go away again",
        "Hello, Dolly",
        "Well, hello, Dolly",
        "It's so nice to have you back where you belong",
        "You're lookin' swell, Dolly",
        "I can tell, Dolly",
        "You're still glowin', you're still crowin'",
        "You're still goin' strong",
        "We feel the room swayin'",
        "While the band's playin'",
        "One of your old favourite songs from way back when",
        "Golly, gee, fellas",
        "Find her a vacant knee, fellas",
        "Dolly'll never go away",
        "Dolly'll never go away",
        "Dolly'll never go away again",
      };

      var sentence = sentences[_random.Next(sentences.Length)];

      Console.WriteLine($"About to 'Say()' ... {sentence}");
      _monitor.Say( sentence, 100, 200 );
      return Task.FromResult(true);
    }

    public Task InitializeAsync(IActionMonitor monitor, CancellationToken token)
    {
      _monitor = monitor ?? throw new ArgumentNullException( nameof(monitor));
      return Task.FromResult(0);
    }
  }
}
